### PR TITLE
[Feature #158761270] Add event update functionality.

### PR DESCRIPTION
**Pull request by parseendavid**
#### What Does This PR Do?
Add feature for updating/editing an event using GraphQL mutation.
#### Description Of Task To Be Completed
- Add mutation for `updateEvent`.
- Add tests for `updateEvent` mutation.

#### Any Background Context You Want To Provide?
- Updating events should be implemented as an `Update Event` feature to non-admin end users on the front end. This allows a creator to change the details of their events.

- I have chosen snapshot testing for [these reasons](http://docs.graphene-python.org/en/latest/testing/#snapshot-testing)

#### How can this be manually tested?
- run `python server/manage.py test server/graphql_schemas/tests`
- run the app, navigate to the `/graphql` endpoint and use the `updateEvent` mutation as shown in the gif below.

#### What are the related pivotal tracker stories
[#158761270](https://www.pivotaltracker.com/story/show/158761270)

#### Screenshot(s)
update_event

### CODE REVIEW

File: server/a_socials/settings/base.py [hunk 1]
```diff
@@ -23,6 +23,7 @@
   )
 )
 
+TEST_RUNNER = 'snapshottest.django.TestRunner'
```
Comment (tosmak16) [inline server/a_socials/settings/base.py]: Well-done @parseendavid  but please let's have this in our dev settings since base settings are for both production and development.

File: server/graphql_schemas/event/schema.py [hunk 2]
```diff
@@ -1,65 +1,107 @@
...
 
     @classmethod
     def mutate_and_get_payload(cls, root, info, **input):
         social_event_id = input.get('social_event_id')
-        social_event = Category.objects.get(id=int(social_event_id))
+        social_event = Category.objects.get(
+            id=from_global_id(social_event_id)[1])
         user = info.context.user
-        googleUser = GoogleUser.objects.get(app_user_id=user.id)
+        googleUser = GoogleUser.objects.first()
```
Comment (tosmak16) [inline server/graphql_schemas/event/schema.py]: @parseendavid Please why the change in line 37?

File: server/graphql_schemas/event/schema.py [hunk 3]
```diff
@@ -1,65 +1,107 @@
...
 
     new_event = graphene.Field(EventNode)
 
     @classmethod
     def mutate_and_get_payload(cls, root, info, **input):
         social_event_id = input.get('social_event_id')
-        social_event = Category.objects.get(id=int(social_event_id))
+        social_event = Category.objects.get(
+            id=from_global_id(social_event_id)[1])
         user = info.context.user
```
Comment (tosmak16) [inline server/graphql_schemas/event/schema.py]: @parseendavid Please why line 36? because `user` is not used due change in line 37.

File: server/graphql_schemas/event/schema.py (hunk 2 above)
Comment (tosmak16) [reply server/graphql_schemas/event/schema.py]: @parseendavid Please why line 36? because `user` is not used any longer due change in line 37.

File: server/graphql_schemas/event/schema.py [hunk 4]
```diff
@@ -1,65 +1,107 @@
...
+        social_event_id = graphene.String(required=False)
 
     new_event = graphene.Field(EventNode)
 
     @classmethod
     def mutate_and_get_payload(cls, root, info, **input):
         social_event_id = input.get('social_event_id')
-        social_event = Category.objects.get(id=int(social_event_id))
+        social_event = Category.objects.get(
+            id=from_global_id(social_event_id)[1])
```
Comment (tosmak16) [inline server/graphql_schemas/event/schema.py]: @parseendavid Please why the change in line 27 and 35? and most importantly what happens if from_global_id returns no value while indexing from index 1?

File: server/graphql_schemas/event/schema.py [hunk 5]
```diff
@@ -1,65 +1,107 @@
...
+        date = graphene.String()
+        time = graphene.String()
+        featured_image = graphene.String()
+        social_event_id = graphene.Int()
+        event_id = graphene.String(required=True)
+
+    updated_event = graphene.Field(EventNode)
+    action_message = graphene.String()
+
+    def update_instance(instance, args, exceptions=['event_id']):
```
Comment (tosmak16) [inline server/graphql_schemas/event/schema.py]: Well done @parseendavid but it will be better if we have the `update_instance` function as a utility function in a utils/utility folder in `graphql_schema` app so that it can be reusable.

File: server/graphql_schemas/event/schema.py [hunk 6]
```diff
@@ -1,65 +1,107 @@
...
+    @classmethod
+    def mutate_and_get_payload(cls, root, info, **input):
+
+        try:
+            user = GoogleUser.objects.get(app_user=info.context.user)
+            event_instance = Event.objects.get(
+                id=from_global_id(input.get('event_id'))[1]
+            )
+            if event_instance.creator != user \
+                    and not info.context.user.is_superuser:
```
Comment (tosmak16) [inline server/graphql_schemas/event/schema.py]: Well done @parseendavid But please why are we checking if the user is not superuser on line 84?

File: server/a_socials/settings/base.py (hunk 1 above)
Comment (parseendavid) [reply server/a_socials/settings/base.py]: I will 👍

File: server/graphql_schemas/event/schema.py (hunk 2 above)
Comment (parseendavid) [reply server/graphql_schemas/event/schema.py]: I forgot to remove that. Thanks for pointing that out. My intention is to use it for the removed line 35. 💯

File: server/graphql_schemas/event/schema.py (hunk 6 above)
Comment (parseendavid) [reply server/graphql_schemas/event/schema.py]: @tosmak16 the intepretation of the line of code is that if a user is not the creator of the event and is not an admin/superuser then the user is not allowed to edit an event.

File: server/graphql_schemas/event/schema.py (hunk 4 above)
Comment (parseendavid) [reply server/graphql_schemas/event/schema.py]: 👍 cool Lemme get into that.

File: server/graphql_schemas/event/schema.py (hunk 5 above)
Comment (parseendavid) [reply server/graphql_schemas/event/schema.py]: 👍

File: server/graphql_schemas/event/schema.py (hunk 4 above)
Comment (parseendavid) [reply server/graphql_schemas/event/schema.py]: For line 27 the id that comes in from the frontend/endpoint is the one global_id which is a string. This was overlooked on the task of creating an event.

File: server/graphql_schemas/event/schema.py (hunk 6 above)
Comment (tosmak16) [reply server/graphql_schemas/event/schema.py]: Okay...Thanks.

File: server/graphql_schemas/event/schema.py [hunk 7]
```diff
@@ -23,43 +25,82 @@ class Input:
...
+class UpdateEvent(relay.ClientIDMutation):
+
+    class Input:
+        title = graphene.String()
+        description = graphene.String()
+        venue = graphene.String()
+        date = graphene.String()
+        time = graphene.String()
+        featured_image = graphene.String()
+        social_event_id = graphene.Int()
```
Comment (tosmak16) [inline server/graphql_schemas/event/schema.py]: well done @parseendavid But line 68 should be graphene.String() like the line  28.

File: server/graphql_schemas/utils/helpers.py [hunk 8]
```diff
@@ -0,0 +1,15 @@
+def update_instance(instance, args, exceptions=['id']):
+    """
+        This function was created to help clean up inputto be used for updating
```
Comment (tosmak16) [inline server/graphql_schemas/utils/helpers.py]: @parseendavid  typo on line 3 `inputto`

File: server/graphql_schemas/event/schema.py (hunk 7 above)
Comment (parseendavid) [reply server/graphql_schemas/event/schema.py]: on it

File: server/graphql_schemas/utils/helpers.py (hunk 8 above)
Comment (parseendavid) [reply server/graphql_schemas/utils/helpers.py]: wow
I really did not catch that.

File: server/graphql_schemas/event/schema.py [hunk 9]
```diff
@@ -23,43 +25,81 @@ class Input:
...
+            event_instance = Event.objects.get(
+                id=from_global_id(input.get('social_event_id'))[1]
+            )
+            if event_instance.creator != user \
+                    and not info.context.user.is_superuser:
+                raise GraphQLError(
+                    "You are not authorized to edit this event.")
+
+            if event_instance:
+                updated_event = update_instance(
```
Comment (tosmak16) [inline server/graphql_schemas/event/schema.py]: @parseendavid This line of code is now failing because of social_event_id is a string and we use from_global_id to get the actual Id value. so there's a need to tweak `update_instance method` or we change the social_event_id value in the input before passing it to `update_instance` helper function.

**Comment by tosmak16** [comment]
Well done @parseendavid :+1: looks good to me.
